SystemInformer/syssccpu.c

```diff
@@ -1822,10 +1822,78 @@ PPH_STRINGREF PhGetHybridProcessorType(
     return NULL;
 }
 #else
+_Success_(return)
+BOOLEAN PhInitializeHybridProcessorTypeCache(
+    _Out_ PPH_LIST *HybridProcessorTypeList
+    )
+{
+    ULONG bufferLength;
+    PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX logicalInformation;
+    PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX processorInfo;
+    PPH_LIST hybridProcessorTypeList = NULL;
+
+    if (!NT_SUCCESS(PhGetSystemLogicalProcessorInformation(RelationProcessorCore, &logicalInformation, &bufferLength)))
+        return FALSE;
+
+    hybridProcessorTypeList = PhCreateList(PhSystemProcessorInformation.NumberOfProcessors);
+
+    for (
+        processorInfo = logicalInformation;
+        (ULONG_PTR)processorInfo < (ULONG_PTR)PTR_ADD_OFFSET(logicalInformation, bufferLength);
+        processorInfo = PTR_ADD_OFFSET(processorInfo, processorInfo->Size)
+        )
+    {
+        //One logical core per bit set in the core mask
+        ULONG_PTR mask = processorInfo->Processor.GroupMask[0].Mask;
+        while (mask)
```

Comment:
Is there a better way to do this?

Reply:
I think this will give you what you want and leverage intrinsic. I believe the intrinsic is supported everywhere?

```c
    ULONG index;

    _BitScanForward(&index, mask);

    mask &= ~(1UL << index);
```

Reply:
Aw, ok, the bit flipping will require less loops because on average the mask become 0 with fewer iterations than scanning all the 64 bits for 1s and 0s.

Reply:
> average the mask become 0 with fewer iterations than scanning all the 64 bits for 1s and 0s

Ah, I see what you're doing here. We use `PhCountBitsUlongPtr` elsewhere which is effectively a popcount, which should be faster than the bit twiddling here?

Reply:
Oh. Nice, akin to https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-rtlnumberofsetbitsulongptr ? (To be frank I never heard about this one either :) First time I'm seeing it)

Reply:
PhCountBitsUlongPtr(), superb, it does exactly what I wanted. I also fixed the previous code since it was only considering the first group mask.